control/noise/handshake.go

```diff
@@ -0,0 +1,430 @@
+// concatenation.
+func (s *symmetricState) MixHash(data []byte) {
+	s.mixer.Reset()
+	s.mixer.Write(s.h[:])
+	s.mixer.Write(data)
+	s.mixer.Sum(s.h[:0]) // TODO: check this actually updates s.h correctly...
+}
+
+// MixDH updates s.ck and s.k with the result of X25519(priv, pub).
+//
+// MixDH corresponds to MixKey(X25519(...))) in the spec. Implementing
+// it as a single function allows for strongly-typed arguments that
+// reduce the risk of error in the caller (e.g. invoking X25519 with
+// two private keys, or two public keys), and thus producing the wrong
+// calculation.
+func (s *symmetricState) MixDH(priv key.Private, pub key.Public) error {
+	// TODO(danderson): check that this operation is correct. The docs
+	// for X25519 say that the 2nd arg must be either Basepoint or the
+	// output of another X25519 call.
+	//
+	// I think this is correct, because pub is the result of a
+	// ScalarBaseMult on the private key, and our private key
+	// generation code clamps keys to avoid low order points. I
+	// believe that makes pub equivalent to the output of
+	// X25519(privateKey, Basepoint), and so the contract is
+	// respected.
+	keyData, err := curve25519.X25519(priv[:], pub[:])
+	if err != nil {
+		return fmt.Errorf("computing X25519: %w", err)
+	}
+
+	r := hkdf.New(newBLAKE2s, keyData, s.ck[:], nil)
+	if _, err := io.ReadFull(r, s.ck[:]); err != nil {
+		return fmt.Errorf("extracting ck: %w", err)
+	}
+	if _, err := io.ReadFull(r, s.k[:]); err != nil {
+		return fmt.Errorf("extracting k: %w", err)
+	}
+	s.n = 0
+	return nil
+}
+
+// EncryptAndHash encrypts plaintext into ciphertext (which must be
+// the correct size to hold the encrypted plaintext) using the current
+// s.k, mixes the ciphertext into s.h, and returns the ciphertext.
+func (s *symmetricState) EncryptAndHash(ciphertext, plaintext []byte) {
+	if s.n == invalidNonce {
+		// Noise in general permits writing "ciphertext" without a
+		// key, but in IK it cannot happen.
+		panic("attempted encryption with uninitialized key")
+	}
+	if len(ciphertext) != len(plaintext)+poly1305.TagSize {
+		panic("ciphertext is wrong size for given plaintext")
+	}
+	aead := newCHP(s.k)
+	var nonce [chp.NonceSize]byte
+	binary.BigEndian.PutUint64(nonce[4:], s.n)
+	s.n++
+	ret := aead.Seal(ciphertext[:0], nonce[:], plaintext, s.h[:])
+	s.MixHash(ret)
+}
+
+// DecryptAndHash decrypts the given ciphertext into plaintext (which
+// must be the correct size to hold the decrypted ciphertext) using
+// the current s.k. If decryption is successful, it mixes the
+// ciphertext into s.h.
+func (s *symmetricState) DecryptAndHash(plaintext, ciphertext []byte) error {
+	if s.n == invalidNonce {
+		// Noise in general permits "ciphertext" without a key, but in
+		// IK it cannot happen.
+		panic("attempted encryption with uninitialized key")
+	}
+	if len(ciphertext) != len(plaintext)+poly1305.TagSize {
+		panic("plaintext is wrong size for given ciphertext")
```

Comment:
I know this is impossible when using the message types, but I am a little worried about DecryptAndHash being later used on attacker-controlled buffers, leading to the panic. In general, when you can return an error, I learned to do that instead and save myself a Denial of Service. :)